File: typescript/Behavioral/16-Mediator/Mediator/DispatchCenter.ts
import { IDispatchMediator } from "./IDispatchMediator";
import { Colleague } from "../Colleague/Colleague";
import { Homeowner } from "../Colleague/Homeowner";
import { Handyman } from "../Colleague/Handyman";
import { Scheduler } from "../Colleague/Scheduler";
import { BillingService } from "../Colleague/BillingService";

export class DispatchCenter implements IDispatchMediator {
  private homeowner?: Homeowner;
  private handyman?: Handyman;
  private scheduler?: Scheduler;
  private billing?: BillingService;

  registerHomeowner(c: Homeowner): void {
    this.homeowner = c;
  }
  registerHandyman(c: Handyman): void {
    this.handyman = c;
  }
  registerScheduler(c: Scheduler): void {
    this.scheduler = c;
  }
  registerBilling(c: BillingService): void {
    this.billing = c;
  }

  notify(
    sender: object,
    eventName: string,
    data: Record<string, unknown> = {},
  ): void {
    const senderName = sender instanceof Colleague ? sender.name : "Unknown";
    console.log(`  [DispatchCenter] Event: '${eventName}' from ${senderName}`);

    switch (eventName) {
      case "JobRequested": {
        const { title, category } = data as {
          title: string;
          category: string;
        };
        const slot = this.scheduler!.bookSlot(category);
        this.handyman!.receiveAssignment(title, slot);
        this.homeowner!.receiveUpdate(
          `Your job '${title}' is booked for ${slot} and assigned to ${this.handyman!.name}.`,
        );
        break;
      }

      case "JobCompleted": {
        const { jobId, handyman } = data as {
          jobId: string;
          handyman: string;
        };
        this.billing!.generateInvoice(jobId, handyman);
        this.homeowner!.receiveUpdate(
          `Job #${jobId} has been completed by ${handyman}. Invoice is on its way.`,
        );
        break;
      }
    }
  }
}

File: csharp/Behavioral/16-Mediator/Mediator/DispatchCenter.cs
public class DispatchCenter : IDispatchMediator
{
    private Homeowner?      _homeowner;
    private Handyman?       _handyman;
    private Scheduler?      _scheduler;
    private BillingService? _billing;

    public void Register(Homeowner      homeowner) => _homeowner = homeowner;
    public void Register(Handyman       handyman)  => _handyman  = handyman;
    public void Register(Scheduler      scheduler) => _scheduler = scheduler;
    public void Register(BillingService billing)   => _billing   = billing;

    public void Notify(object sender, string eventName, object? data = null)
    {
        Console.WriteLine($"  [DispatchCenter] Event: '{eventName}' from {GetName(sender)}");

        switch (eventName)
        {
            case "JobRequested":
                dynamic d = data!;
                var slot = _scheduler!.BookSlot((string)d.Category);
                _handyman!.ReceiveAssignment((string)d.Title, slot);
                _homeowner!.ReceiveUpdate(
                    $"Your job '{d.Title}' is booked for {slot} " +
                    $"and assigned to {_handyman.Name}.");
                break;

            case "JobCompleted":
                dynamic c = data!;
                _billing!.GenerateInvoice((string)c.JobId, (string)c.Handyman);
                _homeowner!.ReceiveUpdate(
                    $"Job #{c.JobId} has been completed by {c.Handyman}. " +
                    $"Invoice is on its way.");
                break;
        }
    }

    private static string GetName(object sender) => sender switch
    {
        Colleague c => c.Name,
        _           => sender.GetType().Name
    };
}
